src/platform_impl/linux/common/xkb_state.rs

```diff
@@ -0,0 +1,694 @@
+                        Some(ffi::xkb_compose_feed_result::XKB_COMPOSE_FEED_ACCEPTED) => (
+                            // Extracting only a single `char` here *should* be fine, assuming that no dead
+                            // key's non-combining variant ever occupies more than one `char`.
+                            Key::Dead(
+                                self.state
+                                    .compose_get_utf8_2()
+                                    .map(|s| s.chars().nth(0).unwrap()),
+                            ),
+                            location,
+                        ),
+                        _ => (key, location),
+                    }
+                }
+                _ => (
+                    self.composed_text()
+                        .unwrap_or_else(|_| self.state.keysym_to_utf8_raw(self.keysym))
+                        .map(Key::Character)
+                        .unwrap_or(key),
+                    location,
+                ),
+            })
+    }
+
+    pub fn key_without_modifiers(&mut self) -> (Key<'static>, KeyLocation) {
+        // This will become a pointer to an array which libxkbcommon owns, so we don't need to deallocate it.
+        let mut keysyms = ptr::null();
+        let keysym_count = unsafe {
+            (XKBH.xkb_keymap_key_get_syms_by_level)(
+                self.state.xkb_keymap,
+                self.keycode + 8,
+                0,
+                0,
+                &mut keysyms,
+            )
+        };
+        let keysym = if keysym_count == 1 {
+            unsafe { *keysyms }
+        } else {
+            0
+        };
+        self.keysym_to_key(keysym)
+            .unwrap_or_else(|(key, location)| {
+                (
+                    self.state
+                        .keysym_to_utf8_raw(keysym)
+                        .map(Key::Character)
+                        .unwrap_or(key),
+                    location,
+                )
+            })
+    }
+
+    fn keysym_to_key(
+        &mut self,
+        keysym: u32,
+    ) -> Result<(Key<'static>, KeyLocation), (Key<'static>, KeyLocation)> {
+        let location = super::keymap::keysym_location(keysym);
+        let key = super::keymap::keysym_to_key(keysym);
+        if matches!(key, Key::Unidentified(_)) {
+            Err((key, location))
+        } else {
+            Ok((key, location))
+        }
+    }
+
+    pub fn text(&mut self) -> Option<&'static str> {
+        self.composed_text()
+            .unwrap_or_else(|_| self.state.keysym_to_utf8_raw(self.keysym))
+    }
+
+    pub fn text_with_all_modifiers(&mut self) -> Option<&'static str> {
+        // TODO: Should Ctrl override any attempts to compose text?
+        //       gnome-terminal agrees, but konsole disagrees.
+        //       Should it be configurable instead?
+        self.composed_text()
+            .unwrap_or_else(|_| self.state.get_utf8_raw(self.keycode))
+    }
```

Comment:
We can come back to this when someone shows up, asking for a specific behaviour, but until then, we should just keep doing what we are doing currently (which is that Ctrl doesn't apply if there's a dead key waiting to be composed).

Reply:
Alright. I've updated the comment to reflect this.